File: 003_Longest_Substring_Without_Repeating_Characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
    	:type s: str
        :rtype: int

        >>> Solution().lengthOfLongestSubstring("abccdabcdaecabedab")
        5
        >>> Solution().lengthOfLongestSubstring("abcabcbb")
        3
        >>> Solution().lengthOfLongestSubstring("pwwkew")
        3
        >>> Solution().lengthOfLongestSubstring("bbbbbb")
        1
        """
        maxLength = 0;
        substr = "";

        for char in s:
        	if char in substr:
        		substr = substr[substr.find(char) + 1 : len(substr)]
        	substr += char
        	maxLength = max(maxLength, len(substr))
        return maxLength

# SOLUTION 1
# class Solution:
#     def lengthOfLongestSubstring(self, s):
#         maxLength, start, dic = 0, 0, {}
        
#         for idx, char in enumerate(s):
#             if char in dic and dic[char] >= start:
#                 start = dic[char] + 1
#             else:
#                 maxLength = max(maxLength, idx - start + 1)
#             dic[char] = idx
            
#         return maxLength

# SOLUTION 2
# class Solution:
#     def lengthOfLongestSubstring(self, s):
#         maxLength, start = 0, 0
        
#         for idx, char in enumerate(s):
#             pos = s.find(char, start, idx)
#             if pos != -1:
#                 start = pos + 1
#             else:
#                 maxLength = max(maxLength, idx - start + 1)
            
#         return maxLength

File: 003_Longest_Substring_Without_Repeating_Characters/Solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let maxLength = 0,
        substr = "";
    for (let i = 0; i < s.length; i ++) {
        let char = s.charAt(i),
            pos = substr.indexOf(char);
        if (pos !== -1) {
            substr = substr.slice(pos + 1, substr.length);
        }
        substr += char;
        maxLength = Math.max(maxLength, substr.length);
    }
    return maxLength;
};

// SOLUTION 1
// var lengthOfLongestSubstring = function(s) {
//     let maxLength = 0,
//         start = 0;
//     for (let i = 0; i < s.length; i ++) {
//         let char = s.charAt(i), 
//             pos = s.indexOf(char, start);
//         if (pos !== -1 && pos < i) {
//             start = pos + 1;
//         } else {
//             maxLength = Math.max(maxLength, i - start + 1);
//         }
//     }
//     return maxLength;
// };